libraries/chain/include/eosio/chain/resource_limits_private.hpp

```diff
@@ -16,13 +16,43 @@ namespace eosio { namespace chain { namespace resource_limits {
 
       template<typename T>
       T operator* (T value, const ratio<T>& r) {
+         EOS_ASSERT(std::numeric_limits<T>::max() / r.numerator >= value, rate_limiting_state_inconsistent, "Usage exceeds maximum value representable after extending for precision");
          return (value * r.numerator) / r.denominator;
       }
 
-      constexpr uint64_t integer_divide_ceil(uint64_t num, uint64_t den ) {
+      template<typename UnsignedIntType>
+      constexpr UnsignedIntType integer_divide_ceil(UnsignedIntType num, UnsignedIntType den ) {
          return (num / den) + ((num % den) > 0 ? 1 : 0);
       }
 
+      /**
+       * Specialization for Signedness matching integer types
+       */
+      template<typename LesserIntType, typename GreaterIntType>
+      constexpr auto downgrade_cast(GreaterIntType val) ->
+         std::enable_if_t<std::is_signed<LesserIntType>::value == std::is_signed<GreaterIntType>::value, LesserIntType>
+      {
```

Comment:
It would be even nicer if all of these templated functions had static asserts to check that the template type is an integer type.